go/pvl/parse.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2016 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package pvl
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+func parse(in string) (pvlT, error) {
+	b := []byte(in)
+	p := pvlT{}
+	p.PvlVersion = -1
+	p.Revision = -1
+	err := json.Unmarshal(b, &p)
+	if p.PvlVersion == -1 {
+		return p, fmt.Errorf("pvl_version required")
+	}
+	if p.Revision == -1 {
+		return p, fmt.Errorf("revision required")
+	}
+	return p, err
+}
+
+type pvlT struct {
+	PvlVersion int `json:"pvl_version"`
+	Revision   int `json:"revision"`
+	// services is a map from service to a list of scripts.
+	// each script is a list of instructions.
+	Services servicesT `json:"services"`
+}
+
+type servicesT struct {
+	Map map[keybase1.ProofType][]scriptT
+}
+
+func (x *servicesT) UnmarshalJSON(b []byte) error {
+	// read as string map
+	m := make(map[string][]scriptT)
+	err := json.Unmarshal(b, &m)
+	if err != nil {
+		return err
+	}
+	// copy to ProofType map
+	x.Map = make(map[keybase1.ProofType][]scriptT)
+	for k, v := range m {
+		t, ok := keybase1.ProofTypeMap[strings.ToUpper(k)]
+		if ok {
+			x.Map[t] = v
+		}
+		// Unrecognized proof types are discarded silently
+		// So that old clients don't break if a new service is added
```

Comment:
One of the downsides of using the json unmarshal system is that there is no context, so all I've got is printf from here. One can only access into this list using valid `ProofType`, so a typo will be met later with `"No entry for service: %v"`.